compiler/optimizer/OMRCFGSimplifier.cpp

```diff
@@ -333,6 +333,292 @@ bool OMR::CFGSimplifier::simplifyInstanceOfTestToCheckcast(bool needToDuplicateT
+//
+//  Or,
+//     if (i>=0 && i < length)
+//        return i;
+//     throw Exception();
+//
+// And tries to simplify them by replacing the ifcmp[lt/ge] blocks with a BNDCHK such that it becomes
+//
+//
+//  Simplification:
+//    BNDCHK (i, length)  ----(exp edge) --> goto
+//      |                                      |
+//    return                                   |
+//      |                                      |
+//    throw Exception() <----------------------
+// Or,
+//
+//  Simplification:
+//    BNDCHK (i, length)  ----(exp edge) --> goto
+//      |                                      |
+//    goto =================                   |
+//      |                   |                  |
+//    throw Exception() <----------------------
+//      |                   |  
+//    return <==============
+//
+// Return "true" if any transformations were made.
+//
+bool OMR::CFGSimplifier::simplifyBoundCheckWithThrowException(bool needToDuplicateTree) 
+   {
+   static char *disableSimplifyBoundCheckWithThrowException = feGetEnv("TR_disableSimplifyBoundCheckWithThrowException");
+   if (disableSimplifyBoundCheckWithThrowException != NULL)
+      return false;
+   if (trace())
+      traceMsg(comp(), "Start simplifyBoundCheckWithThrowException\n");
+   TR::TreeTop * treeTop = getLastRealTreetop(_block);
+   if (!treeTop)
+      return false;
+   TR::Node *compareNode = treeTop->getNode();
+
+   TR::ILOpCodes opCode = compareNode->getOpCodeValue();
+   if (opCode != TR::ificmplt && opCode != TR::ificmpge)
+      return false;
+
+   bool isCompareLT = opCode == TR::ificmplt;
+   bool patternMatched = false;
+
+   TR::Block *throwBlock = NULL;
+   TR::Block *nextCompareBlock = NULL;
+   TR::Node *firstChildBndChk = NULL;
+   TR::Node *secondChildBndChk = NULL;
+
+   if (compareNode->getNumChildren() > 1)
+      {
+      TR::Node *firstChild = compareNode->getFirstChild();
+      TR::Node *secondChild = compareNode->getSecondChild();
+
+      if (firstChild->getOpCodeValue() == TR::iload)
+         {
+         if (secondChild->getOpCodeValue() == TR::iconst && secondChild->getConstValue() == 0)
+            {
+            patternMatched = true;
+            throwBlock = isCompareLT ? compareNode->getBranchDestination()->getNode()->getBlock() : _block->getNextBlock();
+            nextCompareBlock = isCompareLT ? _block->getNextBlock() : compareNode->getBranchDestination()->getNode()->getBlock();
+            firstChildBndChk = compareNode->getFirstChild();
+            }
+         else if (secondChild->getOpCode().isIntegralLoadVar() || secondChild->getOpCode().isLoadIndirect())
+            {
+            patternMatched = true;
+            throwBlock = isCompareLT ? _block->getNextBlock() : compareNode->getBranchDestination()->getNode()->getBlock();
+            nextCompareBlock = isCompareLT ? compareNode->getBranchDestination()->getNode()->getBlock() : _block->getNextBlock();
+            secondChildBndChk = compareNode->getSecondChild();
```

Comment:
we can replace RHS with 'secondChild' since it was already calculated above.

Reply:
I am not sure what you meant

Reply:
I think we already did `compareNode->getSecondChild()`  above the` if`.

Reply:
Done